### PR TITLE
fix: allow Codex auto-routing structured output

### PR DESCRIPTION
**Pull request by nrslib**
## Summary

- make the auto-routing estimator schema valid for Codex strict structured output
- represent optional confidence as a required nullable field and normalize null to an omitted internal value
- fail fast while constructing the estimator when its structured-output schema is invalid, instead of silently selecting `auto.fallback`
- keep the router schema within the raw structured-output subset shared by Codex and Claude, with wire-lossless validation
- add Codex regression coverage and a Claude integration test through the Agent SDK `query()` boundary

## Root cause

The estimator schema declared `confidence` in `properties` but omitted it from `required`. Codex rejects that strict output schema, so every estimator call was handled as a failure and selected the configured `auto.fallback` candidate.

No additional user configuration is required. `confidence` is part of TAKT's internal estimator response contract, not an `auto_routing` configuration field.

## Failure policy

- deterministic schema/configuration mismatches fail synchronously before the provider call
- runtime provider failures and invalid model responses continue to use the configured fallback

## Verification

- `npm run build`
- `npm run lint`
- type-contracts
- 170 related unit tests
- Claude Agent SDK `query()` integration test
- `git diff --check`
- independent Codex sol review using `$coding`: APPROVE, 0 findings

A local full `npm test` run used Node 22.22.1 and failed only in Node 24.15+ `node:sqlite` API coverage (`DatabaseSync.enableDefensive`) and a sandbox-restricted Unix socket test. The affected auto-routing tests passed, and CI runs with the repository-supported runtime.

## Summary by CodeRabbit

* **機能改善**
  * ワーク要件の推定結果に信頼度（confidence / number または null）を追加し、null の場合は返却に含めない仕様にしました。
  * 返却形式の厳密なスキーマ検証を強化し、不正な structured output や schema 不整合を適切に拒否するよう調整しました。
* **テスト**
  * confidence の省略／null／無効値の扱い、および厳格スキーマ条件での受理・拒否を追加・更新しました。
  * Claude 経由の統合テストを追加し、期待する出力形式を確認しました。

### CODE REVIEW

File: src/agents/auto-routing-usecase.ts [hunk 1]
```diff
@@ -107,13 +107,13 @@ function parseEstimate(parsed: unknown): WorkRequirementEstimate {
...
   validateRoutingReasonCodes(value.reason_codes);
-  if (value.confidence !== undefined && typeof value.confidence !== 'number') {
+  if (value.confidence !== undefined && value.confidence !== null && typeof value.confidence !== 'number') {
     throw new Error('Auto routing estimator response has invalid confidence');
   }
   return {
     requiredTier: value.required_tier,
     reasonCodes: [...value.reason_codes],
-    ...(value.confidence !== undefined ? { confidence: value.confidence } : {}),
+    ...(typeof value.confidence === 'number' ? { confidence: value.confidence } : {}),
```
[the remark this answers is not in the document]
Comment (nrslib) [reply src/agents/auto-routing-usecase.ts:113]: 再検証の結果、この指摘は方針不一致として本番コードは変更しません。`OUTPUT_SCHEMA.required` は native structured-output provider の生成制約ですが、`parseEstimate` は `structuredOutput` と通常の JSON content を統合する境界です。Cursor/Kiro/Copilot は `outputSchema` 非対応で、内部契約も `confidence?: number` のため、欠落拒否は既存provider経路を破壊します。意図を固定するため、commit 34c5ba17 で Cursor の `confidence` 欠落JSONを受理する回帰テストを追加しました。数値は保持、null/欠落は省略、不正型は引き続き拒否します。独立Codex sol再レビューは APPROVE（finding 0件）、関連39テストとtype-contractsも成功しています。

File: src/core/workflow/engine/structured-output-schema-validator.ts [hunk 2]
```diff
@@ -105,6 +145,370 @@ function getValidator(schema: Record<string, unknown>): ValidateFunction {
...
+    }
+  }
+
+  if (
+    schemaTypes?.includes('string')
+    && typeof schema.format === 'string'
+    && !SUPPORTED_RAW_STRING_FORMATS.has(schema.format)
+  ) {
+    throw strictSchemaError(`${path} uses unsupported format ${schema.format}`);
+  }
```
[the remark this answers is not in the document]
Comment (nrslib) [reply src/core/workflow/engine/structured-output-schema-validator.ts]: 対応しました。commit b302e656 で  がown propertyの場合に非string値を明示的な  で拒否し、allowlist外formatの既存拒否も維持しました。具体的なエラーメッセージを検証する回帰テストも追加しています。

File: src/core/workflow/engine/structured-output-schema-validator.ts [hunk 3]
```diff
@@ -105,6 +145,370 @@ function getValidator(schema: Record<string, unknown>): ValidateFunction {
...
+    (property): property is string => (
+      typeof property === 'string'
+      && !Object.hasOwn(properties, property)
+    ),
+  );
+  if (unknownProperties.length > 0) {
+    throw strictSchemaError(
+      `${path} must only require declared properties (unknown: ${unknownProperties.join(', ')})`,
+    );
+  }
```
[the remark this answers is not in the document]
Comment (nrslib) [reply src/core/workflow/engine/structured-output-schema-validator.ts:385]: 対応しました。commit b302e656 で  非plain object、 非array、 の非string要素をfallbackせず明示的なstrict schema errorとして拒否しました。3つの負系と正常系テストを追加し、format非stringの明示拒否も同commitに含めています。

File: src/core/workflow/engine/structured-output-schema-validator.ts [hunk 4]
```diff
@@ -105,6 +145,370 @@ function getValidator(schema: Record<string, unknown>): ValidateFunction {
...
+          nested.schema,
+          nested.path,
+          new Set([...activeSchemas, nested.schema]),
+        );
+      }
+    }
+  };
+
+  visit(rootSchema, '$', new Set([rootSchema]));
+}
```
[the remark this answers is not in the document]
Comment (nrslib) [reply src/core/workflow/engine/structured-output-schema-validator.ts:508]: 再検証の結果、memo化は追加しません。 の本番callsiteは auto-router のstatic  をfactory生成時に1回検証する箇所だけで、外部workflow schema経路では使われていません。またAjv compileは既存WeakMapでcache済みです。この利用範囲では追加の走査cacheは効果がなく、循環判定を複雑化するため no_issue_after_verification としました。独立Codex solレビューも同判定です。

File: src/__tests__/structured-output-schema-validator.test.ts [hunk 5]
```diff
@@ -38,4 +79,684 @@ describe('structured output schema validator', () => {
...
+            },
+            required: [],
+            additionalProperties: false,
+          },
+        },
+        required: ['nested'],
+        additionalProperties: false,
+      },
+      expectedMessage: '$.nested',
+    },
```
[the remark this answers is not in the document]
Comment (nrslib) [reply src/__tests__/structured-output-schema-validator.test.ts:115]: 対応しました。commit b302e656 で expectedMessage を  まで具体化し、意図した拒否理由を固定しました。

File: src/core/workflow/engine/structured-output-schema-validator.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (nrslib) [reply src/core/workflow/engine/structured-output-schema-validator.ts]: 対応しました。commit b302e656 で format がown propertyの場合に非string値を明示的な StructuredOutputSchemaError で拒否し、allowlist外formatの既存拒否も維持しました。具体的なエラーメッセージを検証する回帰テストも追加しています。

File: src/core/workflow/engine/structured-output-schema-validator.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (nrslib) [reply src/core/workflow/engine/structured-output-schema-validator.ts:508]: 再検証の結果、memo化は追加しません。assertStrictStructuredOutputSchema の本番callsiteは auto-router のstatic OUTPUT_SCHEMA をfactory生成時に1回検証する箇所だけで、外部workflow schema経路では使われていません。またAjv compileは既存WeakMapでcache済みです。この利用範囲では追加の走査cacheは効果がなく、循環判定を複雑化するため no_issue_after_verification としました。独立Codex solレビューも同判定です。

File: src/core/workflow/engine/structured-output-schema-validator.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nrslib) [reply src/core/workflow/engine/structured-output-schema-validator.ts:385]: 対応しました。commit b302e656 で properties 非plain object、required 非array、required の非string要素をfallbackせず明示的なstrict schema errorとして拒否しました。3つの負系と正常系テストを追加し、format非stringの明示拒否も同commitに含めています。

File: src/__tests__/structured-output-schema-validator.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (nrslib) [reply src/__tests__/structured-output-schema-validator.test.ts:115]: 対応しました。commit b302e656 で expectedMessage を $.nested must list every property in required (missing: value) まで具体化し、意図した拒否理由を固定しました。

File: src/__tests__/work-requirement-estimator.test.ts [hunk 6]
```diff
@@ -2,11 +2,21 @@ import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
...
+}));
 
 vi.mock('../agents/runner.js', () => ({
   runAgent: vi.fn(),
 }));
 
+vi.mock('../core/workflow/engine/structured-output-schema-validator.js', async (importOriginal) => ({
+  ...await importOriginal<typeof import('../core/workflow/engine/structured-output-schema-validator.js')>(),
+  assertStrictStructuredOutputSchema,
+}));
```
[the remark this answers is not in the document]
Comment (nrslib) [reply src/__tests__/work-requirement-estimator.test.ts:18]: 再検証の結果、この指摘は追加変更不要です。src/__tests__/work-requirement-estimator-claude.integration.test.ts は validator を mock せず、実 createWorkRequirementEstimator から Claude Agent SDK の query 境界まで通しています。したがって実物の OUTPUT_SCHEMA が assertStrictStructuredOutputSchema を通らなければ provider 呼び出し前にこのITが失敗します。独立Codex solレビューでも同経路を確認し、no_issue_after_verification と判定しました。
